### PR TITLE
Floating IPs

### PR DESCRIPTION
**Pull request by tejasmanohar**
v0 of Floating IPs. They're implemented as machine rules, which are at the label/service, not container-level, so that's a bit awkward since 1 IP : 1 container, but I think we should fix this in a refactor of placement rules (and maybe under the hood-- e.g. assign a rule of ["eip-1", "eip-2"] to a label with 2 containers).

### CODE REVIEW

File: cluster/provider/amazon_test.go [hunk 1]
```diff
@@ -475,3 +476,70 @@ func TestStop(t *testing.T) {
 		},
 	)
 }
+
+func TestAssociateFloatingIP(t *testing.T) {
```
Comment (tejasmanohar) [inline cluster/provider/amazon_test.go]: do we prefer to just test public functions (e.g. Boot)? since there are multiple calls to the same AWS APIs and they all accept pointer arguments (making it hard to specify specific input (hence the use of mock.Anything)), I find it harder to read but can figure something out if that's preferred // cc @kklin

**Comment by tejasmanohar** [comment]
@kklin is reviewing.

**Comment by tejasmanohar** [comment]
oops, need to update minion tests

**Comment by tejasmanohar** [comment]
Updated spec examples & code to allow users to enter their floating IP's (GCE NAT / AWS Elastic IP) public address instead of ID as per @ejj's suggestion.

File: cluster/cluster.go [hunk 2]
```diff
@@ -283,6 +286,8 @@ func syncDB(cloudMachines []provider.Machine, dbMachines []db.Machine) (
 			return -1
 		case m.Region != "" && dbm.Region != m.Region:
 			return -1
+		case m.FloatingIP != "" && dbm.FloatingIP != m.FloatingIP:
```
Comment (kklin) [inline cluster/cluster.go]: Can we just do `dbm.FloatingIP != m.FloatingIP`? I think the `m.Region != "" && ...` etc is leftover from before regions and sizes were implemented in all providers.

Also, this won't disassociate floating IP's right? What if you want to switch the IP on a machine for whatever reason?

File: stitch/bindings.js [hunk 3]
```diff
@@ -420,6 +422,9 @@ function MachineRule(exclusive, optionalArgs) {
     if (optionalArgs.region) {
         this.region = optionalArgs.region;
     }
+    if (optionalArgs.floatingIp) {
+      this.floatingIp = optionalArgs.floatingIp;
```
Comment (kklin) [inline stitch/bindings.js]: I thought we were going to error if someone applies a floatingIP placement rule with a service that has multiple containers, and accepts connections from `publicInternet`?

File: cluster/provider/awsSpot.go [hunk 4]
```diff
@@ -385,6 +411,84 @@ func (clst *amazonCluster) ChooseSize(ram stitch.Range, cpu stitch.Range,
 	return pickBestSize(constants.AwsDescriptions, ram, cpu, maxPrice)
 }
 
+func (clst amazonCluster) associateFloatingIPs(ips []floatingIP) error {
```
Comment (kklin) [inline cluster/provider/awsSpot.go]: Not saying we should do that before this commit, btw. I do think the code can be simplified, though. Will think about it later.

File: stitch/bindings.js [hunk 5]
```diff
@@ -262,7 +262,8 @@ Service.prototype.getQuiltPlacements = function() {
             otherLabel: placement.otherLabel || "",
             provider: placement.provider || "",
             size: placement.size || "",
-            region: placement.region || ""
+            region: placement.region || "",
+            floatingIp: placement.floatingIp || ""
```
Comment (tejasmanohar) [inline stitch/bindings.js]: i've always seen acronyms treated as words in JS camelCase (unlike Go), but I don't feel strongly. Let me know if you think we should still change this.

File: cluster/cluster.go (hunk 2 above)
Comment (tejasmanohar) [inline cluster/cluster.go]: yep, i think you're right. changed.

if you're also concerned about disassociating IPs on AWS, I already tested that we don't need to add anything to Stop(). next associate call just unassigns & re-assigns.

File: cluster/provider/awsSpot.go (hunk 4 above)
Comment (tejasmanohar) [inline cluster/provider/awsSpot.go]: agreed. gave it a 2nd look & made some changes. let me know what you think!

File: specs/example-ip-gce.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+// https://cloud.google.com/compute/docs/configure-instance-ip-addresses#reserve_new_static
```
Comment (tejasmanohar) [inline specs/example-ip-gce.js]: done

File: stitch/bindings.js (hunk 3 above)
Comment (tejasmanohar) [inline stitch/bindings.js]: added

File: stitch/bindings.js (hunk 5 above)
Comment (kklin) [inline stitch/bindings.js]: Ah, didn't realize that was the standard in Javascript. Let's leave it camelcase then.

File: cluster/provider/awsSpot.go [hunk 7]
```diff
@@ -455,6 +541,75 @@ OuterLoop:
 	return errors.New("timed out")
 }
 
+// getInstanceIDs builds a map of spot IDs to instance IDs from spotIDs
+// for a given region.
+func (clst amazonCluster) getInstanceIDs(region string,
```
Comment (kklin) [inline cluster/provider/awsSpot.go]: I think we could simplify `getInstanceIDs` and `waitOnInstances`. They're pretty similar.

What if we defined a (non-blocking) `getInstances([]awsID) (map[awsID]*instance, error)`? `associateFloatingIPs` would then wait until all instances were `!= nil && running`.. We might also define some sort of `waiter` function, not sure.

What do you think?

Comment (tejasmanohar) [inline cluster/provider/awsSpot.go]: I started on this path, but I found the control flow more awkward. Since all AWS actions require grouping by region, I think it's much cleaner to group by region at the very beginning of the `associateFloatingIPs` function. Otherwise, you group aws IDs on every action (e.g. after getting instances from `getInstances`, you have to group results again to associate floating IPs-- which is also weird with the waiter function)

Comment (tejasmanohar) [inline cluster/provider/awsSpot.go]: To illustrate, here's a (rough, definitely non-functional) version that takes that approach-- https://gist.github.com/tejasmanohar/50d26d768f09b8ae7536920991b59abb

Comment (kklin) [inline cluster/provider/awsSpot.go]: I'm not saying we get rid of grouping by region first, but that we just define a single `getInstances` function rather than having both `getInstanceIDs` and `waitOnInstances`.

```
func wait(pred func() bool) error

for region, floatingIPs := range toAssociate {
    wait(func() bool {
        for _, inst := range getInstances(region, spotIDs) {
            if inst == nil || inst.state != running {
                return false
            }
        }
        return true
    }

    instanceMap := getInstances(region, spotIDs)
    for _, ip := range floatingIPs {
        associate(instanceMap[ip.spotID].instanceID, ip.address)
    }
}
```

Comment (tejasmanohar) [inline cluster/provider/awsSpot.go]: ah ok, ^ that would mean we retry all spot IDs everytime... but i guess that's fine to start.

Comment (tejasmanohar) [inline cluster/provider/awsSpot.go]: anyways, i'll do that! but probably without the wait() fn and just a loop

Comment (kklin) [inline cluster/provider/awsSpot.go]: It's the same runtime because we're still iterating over all the current SpotIDRequests right? Either way, I'd happily trade more obviously correct code for something less efficient until it shows up as a bottleneck.

btw, the reason why I proposed the `wait()` function is because of `clst.wait`. Not sure how much cleaner the function would make that code, though.

**Comment by tejasmanohar** [comment]
Ready for 2nd review @kklin

File: stitch/bindings.js (hunk 5 above)
Comment (tejasmanohar) [inline stitch/bindings.js]: just how i've seen it & how we do things at work at least, i only starting writing `floatingIP` with the capital `P` because of gometalinter :P

File: cluster/cluster.go [hunk 8]
```diff
@@ -283,6 +286,8 @@ func syncDB(cloudMachines []provider.Machine, dbMachines []db.Machine) (
 			return -1
 		case m.Region != "" && dbm.Region != m.Region:
```
Comment (tejasmanohar) [inline cluster/cluster.go]: saw loops of `Attempting to Boot machines`, `Attempting to halt machines` after removing the != "" checks so requires more investigation (probably outside of this PR?)

Comment (tejasmanohar) [inline cluster/cluster.go]: nevermind, that seems unrelated. strange enough, after attempting to boot/halt a bunch, it still prints the same spot instance request IDs

Comment (kklin) [inline cluster/cluster.go]: Why do you say it's unrelated? Could you post the log? But yeah, no need to spend too much time on it, could you just file a card on the issue?

Also, it's possible that it created and destroyed other spot requests, and kept the same ones in the DB because of our join score.

File: stitch/stitch_test.go [hunk 9]
```diff
@@ -73,6 +73,20 @@ func TestMachine(t *testing.T) {
 			},
 		},
 	)
+
+	checkMachines(t, `var baseMachine = new Machine({
+	  provider: "Amazon",
+	  floatingIp: "eipalloc-123"
```
Comment (kklin) [inline stitch/stitch_test.go]: Same here re making the test use the IP address

File: cluster/provider/awsSpot.go [hunk 10]
```diff
@@ -385,6 +427,115 @@ func (clst *amazonCluster) ChooseSize(ram stitch.Range, cpu stitch.Range,
...
+			}
+
+			for _, inst := range instances {
+				if inst == nil ||
+					*inst.State.Name != ec2.InstanceStateNameRunning {
+					return true
+				}
+			}
+			return false
+		}, 100, 10*time.Second)
```
Comment (kklin) [inline cluster/provider/awsSpot.go]: I think it'd be much more intuitive if `Wait` took a timeout rather than number of times to try.

File: cluster/provider/awsSpot.go [hunk 11]
```diff
@@ -385,6 +427,115 @@ func (clst *amazonCluster) ChooseSize(ram stitch.Range, cpu stitch.Range,
...
+		if err != nil {
+			return err
+		}
+
+		for _, ip := range floatingIPs {
+			_, err := session.AssociateAddress(&ec2.AssociateAddressInput{
+				InstanceId: aws.String(
+					*instances[ip.spotID].InstanceId,
```
Comment (kklin) [inline cluster/provider/awsSpot.go]: Is this already a *string? If so, we can just assign it directly InstanceId

File: stitch/stitch_test.go (hunk 9 above)
Comment (tejasmanohar) [inline stitch/stitch_test.go]: fixed

File: minion/scheduler/master_test.go [hunk 12]
```diff
@@ -75,6 +75,7 @@ func TestCleanup(t *testing.T) {
 			Exclusive:   true,
 			TargetLabel: "1",
 			Region:      "Region1",
+			FloatingIP:  "eipalloc-123",
```
Comment (tejasmanohar) [inline minion/scheduler/master_test.go]: fixed

File: cluster/provider/awsSpot.go (hunk 11 above)
Comment (tejasmanohar) [inline cluster/provider/awsSpot.go]: fixed

File: cluster/provider/awsSpot.go (hunk 10 above)
Comment (tejasmanohar) [inline cluster/provider/awsSpot.go]: fixed, set 15m arbitrarily for now

**Comment by tejasmanohar** [comment]
Ready for review by @ejj.

File: util/util.go [hunk 13]
```diff
@@ -144,3 +146,25 @@ func StrStrMapEqual(x, y map[string]string) bool {
 	}
 	return true
 }
+
+// Wait retries a function up to X times until it returns false with Y constant
+// backoff, erroring if it never completes.
+func Wait(f func() bool, backoff time.Duration, timeout time.Duration) error {
```
Comment (tejasmanohar) [inline util/util.go]: done

**Comment by tejasmanohar** [comment]
@ejj Rebased & ready for review again. Thanks!

**Comment by ejj** [comment]
Eh, sorry @tejasmanohar could you rebase again?  Seems to be conflicting with some new changes.

File: cluster/provider/provider.go [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+import (
+	"fmt"
+
+	"github.com/NetSys/quilt/db"
+	"github.com/NetSys/quilt/stitch"
+)
+
+// Machine represents an instance of a machine booted by a Provider.
+type Machine struct {
```
Comment (ejj) [inline cluster/provider/provider.go]: This looks like a rebasing error.  The machine struct is already in the machine package.

File: cluster/provider/provider.go [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+	FloatingIP string
+	Size       string
+	DiskSize   int
+	SSHKeys    []string
+	Provider   db.Provider
+	Region     string
+}
+
+// ACL represents allowed traffic to a machine.
+type ACL struct {
```
Comment (ejj) [inline cluster/provider/provider.go]: Same with the ACL struct

Comment (ejj) [inline cluster/provider/provider.go]: Yeah most of this file should be deleted.

File: engine/engine.go [hunk 16]
```diff
@@ -155,6 +156,9 @@ func machineTxn(view db.Database, stitch stitch.Stitch) error {
 			return -1
 		case dbMachine.Size != "" && stitchMachine.Size != dbMachine.Size:
 			return -1
+		case dbMachine.FloatingIP != "" &&
```
Comment (ejj) [inline engine/engine.go]: Could you join it and just change the floating IP, or do you have to delete the machine and re-create it to set it's floating IP

File: engine/engine.go [hunk 17]
```diff
@@ -177,7 +181,6 @@ func machineTxn(view db.Database, stitch stitch.Stitch) error {
 
 	for _, bootSet := range bootList {
 		bootSet := bootSet.(db.Machine)
-
```
Comment (ejj) [inline engine/engine.go]: Useless diff

File: stitch/bindings.js [hunk 18]
```diff
@@ -91,13 +91,24 @@ Deployment.prototype.vet = function() {
...
             var otherLabel = plcm.otherLabel;
             if (otherLabel !== undefined && !labelMap[otherLabel]) {
                 throw service.name + " has a placement in terms of an " +
                     "undeployed service: " + otherLabel;
             }
         });
+
+        if (hasFloatingIp && service.incomingPublic.length
+            && service.containers.length > 1) {
+            throw service.name + " has a floating IP and multiple containers. " +
```
Comment (ejj) [inline stitch/bindings.js]: File a trello card on this, or is it fixed in this series?

Comment (tejasmanohar) [inline stitch/bindings.js]: I think there's a card about reworking placement rules to be in terms of containers, not labels. Will note there.

**Comment by tejasmanohar** [comment]
When I rebased again, cluster/provider/* went away so guess it was incomplete before.

File: stitch/bindings.js (hunk 18 above)
Comment (tejasmanohar) [inline stitch/bindings.js]: Opened card

File: engine/engine.go (hunk 17 above)
Comment (tejasmanohar) [inline engine/engine.go]: removed

File: engine/engine.go (hunk 16 above)
Comment (tejasmanohar) [inline engine/engine.go]: yep, seems possible but noted as low-priority. should we do that in this iteration? i'd add set/unsetFloatingIP methods to cluster.provider{}

**Comment by tejasmanohar** [comment]
Ready for next round of review @ejj (maybe @kklin too since the last commit is a large refactor). Addressed https://github.com/NetSys/quilt/pull/587#discussion_r90688737.

File: cluster/cluster.go [hunk 19]
```diff
@@ -36,6 +37,23 @@ type cluster struct {
 	providers map[db.Provider]provider
 }
 
+type syncSet struct {
```
Comment (tejasmanohar) [inline cluster/cluster.go]: some of this probably "should" have been in its own series... but it was really hard to read the code when adding even more arguments and return values (especially in tests)

File: cluster/provider/provider.go (hunk 14 above)
Comment (ejj) [inline cluster/provider/provider.go]: ping ...

**Comment by tejasmanohar** [comment]
Rebased but for some reason, when removing an IP, the cluster now removes it properly... and then tries to re-add it so one of the tests started failing

```
time="2016-12-17T17:48:51-08:00" level=info msg="Attempt to remove floating IPs machines." count=1
time="2016-12-17T17:48:51-08:00" level=info msg="Successfully removed floating IPs from machines"
time="2016-12-17T17:48:51-08:00" level=info msg="db.Machine:\n\tMachine-2{Master, FakeAmazon  m4.large, 1, FloatingIP=another.ip}\n\tMachine-4{Master, FakeVagrant  vagrant.large, 1}\n\tMachine-5{Master, FakeAmazon  m4.large, 3, FloatingIP=ip}"
time="2016-12-17T17:48:51-08:00" level=info msg="db.Machine:\n\tMachine-2{Master, FakeAmazon  m4.large, 1, FloatingIP=another.ip}\n\tMachine-4{Master, FakeVagrant  vagrant.large, 1}\n\tMachine-5{Master, FakeAmazon  m4.large, 3, FloatingIP=ip}"
time="2016-12-17T17:48:51-08:00" level=error msg="Failed to get ACLs" error="no ACL rows found"
time="2016-12-17T17:48:51-08:00" level=info msg="Attempt to assign floating IPs machines." count=1
time="2016-12-17T17:48:51-08:00" level=info msg="Successfully assigned floating IPs to machines"
```

I think it has something to do with join being called twice in a loop. Will keep looking but wanted to post an update here in case someone has any ideas what could've changed to cause this

**Comment by kklin** [comment]
@tejasmanohar I took a look, looks like it's because of this conditional https://github.com/NetSys/quilt/pull/587/files#diff-f2792de55ee0d606be19884f969036cdR261. We don't update the DB to reflect that the floating IP was removed.

**Comment by tejasmanohar** [comment]
derp :( - that makes sense.

**Comment by tejasmanohar** [comment]
@LuiseV Ready for review!

File: specs/example-ip.js [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+
+var baseMachine = new Machine({
+  provider: "Amazon",
+  size: "m4.large",
+  region: "us-west-2",
+  sshKeys: githubKeys("ejj")
+});
+deployment.deploy(baseMachine.asMaster());
+
+baseMachine.floatingIp = floatingIp;
```
Comment (LuiseV) [inline specs/example-ip.js]: We should probably also think about how we'd create multiple machines with floating IPs. If we `replicate` the machines, the current solution would involve some ugly array indexing, I think... There should be a better solution, but I don't think you should care about that now.

File: cluster/amazon/amazon.go [hunk 21]
```diff
@@ -117,14 +126,28 @@ func (clst Cluster) Boot(bootSet []machine.Machine) error {
 			awsIDs = append(awsIDs, awsID{
 				spotID: *request.SpotInstanceRequestId,
 				region: br.region})
+
+			if br.floatingIP != "" {
+				floatingIPs = append(floatingIPs, floatingIP{
+					awsID: awsID{
+						spotID: *request.SpotInstanceRequestId,
```
Comment (tejasmanohar) [inline cluster/amazon/amazon.go]: not sure i follow... https://play.golang.org/p/UcJ_gtIbGp

File: cluster/vagrant/vagrant.go [hunk 22]
```diff
@@ -1,6 +1,7 @@
 package vagrant
```
Comment (tejasmanohar) [inline cluster/vagrant/vagrant.go]: IMO, no. e58bc39 introduces new methods on the `provider` interface to make "Floating IPs join()'able in-place", and this just makes Vagrant _satisfy_ the interface. Wouldn't want an in-between commit with compile errors on master :P

File: cluster/amazon/amazon.go [hunk 23]
```diff
@@ -314,6 +352,113 @@ func (clst Cluster) getClient(region string) client {
...
+			}
+
+			for _, inst := range instances {
+				if inst == nil ||
+					*inst.State.Name != ec2.InstanceStateNameRunning {
+					return false
+				}
+			}
+			return true
+		}, 10*time.Second, 15*time.Minute)
```
Comment (tejasmanohar) [inline cluster/amazon/amazon.go]: kept it ~equal to what the timeout was for other stuff in this package, but I agree. what do you think is more reasonable? idk if things slow down as you boot more machines in parallel in one account etc (though I wouldn't expect that)

File: cluster/provider/provider.go (hunk 14 above)
Comment (ejj) [inline cluster/provider/provider.go]: Right, I'm saying the cluster machines moved to cluster/machine/machine.go.

@tejas, are you going to fix the rebasing error?

File: cluster/provider/provider.go [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+	case "Vagrant":
+	default:
+		panic(fmt.Sprintf("Unknown Cloud Provider: %s", m.Provider))
+	}
+
+	m.Region = region
+	return m
+}
+
+func resolveString(ptr *string) string {
```
Comment (tejasmanohar) [inline cluster/provider/provider.go]: (ignoring)

File: cluster/provider/provider.go [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+// they don't particularly care where a system is placed.
+func DefaultRegion(m db.Machine) db.Machine {
+	if m.Region != "" {
+		return m
+	}
+
+	region := ""
+	switch m.Provider {
+	case "Amazon":
+		region = "us-west-1"
```
Comment (tejasmanohar) [inline cluster/provider/provider.go]: (ignoring)

File: cluster/provider/provider.go [hunk 26]
```diff
@@ -1,121 +0,0 @@
-package provider
```
Comment (tejasmanohar) [inline cluster/provider/provider.go]: (ignoring)

File: cluster/amazon/amazon.go [hunk 27]
```diff
@@ -314,6 +352,113 @@ func (clst Cluster) getClient(region string) client {
 	return clst.clients[region]
 }
 
+func (clst Cluster) associateFloatingIPs(requests []floatingIP) error {
+	if len(requests) == 0 {
+		return nil
+	}
+
+	// toAssociate is a map of region to spot ID to floating IP.
+	toAssociate := map[string]map[string]floatingIP{}
```
Comment (tejasmanohar) [inline cluster/amazon/amazon.go]: (outdated commit)

File: specs/example-ip.js (hunk 20 above)
Comment (tejasmanohar) [inline specs/example-ip.js]: nope, no reason!

and yes, I agree. will think on that

File: specs/example-ip.js [hunk 28]
```diff
@@ -0,0 +1,24 @@
...
+publicInternet.connect(80, webTier);
+
+webTier.place(new MachineRule(false, {floatingIp: floatingIp}));
+deployment.deploy(webTier);
+
+var baseMachine = new Machine({
+  provider: "Amazon",
+  size: "m4.large",
+  region: "us-west-2",
+  sshKeys: githubKeys("ejj")
```
Comment (tejasmanohar) [inline specs/example-ip.js]: sure!

**Comment by tejasmanohar** [comment]
Updated @LuiseV

File: cluster/provider/provider.go (hunk 14 above)
Comment (tejasmanohar) [inline cluster/provider/provider.go]: weird, yes

File: cluster/provider/provider.go (hunk 26 above)
Comment (tejasmanohar) [inline cluster/provider/provider.go]: fixed rebase error

File: cluster/amazon/amazon.go [hunk 29]
```diff
@@ -210,6 +233,17 @@ func (clst Cluster) List() ([]machine.Machine, error) {
...
 		}
 
+		addrResp, err := client.DescribeAddresses(nil)
+		if err != nil {
+			return nil, err
+		}
+		ipMap := map[string]*ec2.Address{}
+		for _, ip := range addrResp.Addresses {
+			if ip.InstanceId != nil {
+				ipMap[*ip.InstanceId] = ip
```
Comment (LuiseV) [inline cluster/amazon/amazon.go]: Oh, wait, my bad. You are dereferencing the member, not the struct.

File: cluster/amazon/amazon.go (hunk 23 above)
Comment (LuiseV) [inline cluster/amazon/amazon.go]: Yeah, I'd say something between 5 and 10 minutes.

**Comment by LuiseV** [comment]
Ready for @ejj

File: specs/example-ip.js [hunk 30]
```diff
@@ -0,0 +1,23 @@
...
+webTier.place(new MachineRule(false, {floatingIp: floatingIp}));
+deployment.deploy(webTier);
+
+var baseMachine = new Machine({
+  provider: "Amazon",
+  size: "m4.large",
+  region: "us-west-2",
+  // AWS: http://docs.aws.amazon.com/cli/latest/reference/ec2/allocate-address.html
+  // Google: https://cloud.google.com/compute/docs/configure-instance-ip-addresses#reserve_new_s
+  floatingIp: "xxx.xxx.xxx.xxx (CHANGE ME)",
```
Comment (kklin) [inline specs/example-ip.js]: I don't think setting it here works:  this will cause the master and worker to both get the same floating IP right?

Comment (kklin) [inline specs/example-ip.js]: Also, looks like the Google link is cutoff.

Comment (LuiseV) [inline specs/example-ip.js]: Good point about setting the floating IP. Sorry for missing that.
I guess you could do something like
```javascript
var worker = baseMachine.asWorker();
worker.floatingIp = "xxx.xxx.xxx.xxx (CHANGE ME)";
deployment.deploy(worker);```
Not sure if you saw the above comment on this, @kklin, but we should probably think about a nice way to assign floating IPs to many different VMs. I imagine it won't be super clean with `replicate()`.

Comment (tejasmanohar) [inline specs/example-ip.js]: eek, yeah- that's why I set it separately earlier :( (changed back)

Comment (kklin) [inline specs/example-ip.js]: @tejasmanohar I think it deserves a comment too. Code that looks buggy is basically as bad as buggy code IMO :)

File: cluster/amazon/amazon.go (hunk 29 above)
Comment (tejasmanohar) [inline cluster/amazon/amazon.go]: ah, i see where you got confused

**Comment by ejj** [comment]
Is there a rebase error in this patch `util: Refactor quilt-tester and awsSpot to use util.WaitFor`?  I see you delete the code and tests, but I don't see you add it back
